plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,324 @@
+                searchResults -> {
+                    ImmutableSet.Builder<String> distinguishedNames = ImmutableSet.builder();
+                    while (searchResults.hasMore()) {
+                        distinguishedNames.add(searchResults.next().getNameInNamespace());
+                    }
+                    return distinguishedNames.build();
+                });
+    }
+
+    public Set<String> lookupUserAttributes(String searchBase,
+            String searchFilter,
+            String attributes,
+            String contextUserDistinguishedName,
+            String contextPassword)
+            throws NamingException
+    {
+        return client.executeLdapQuery(
+                contextUserDistinguishedName,
+                contextPassword,
+                new LdapQuery.LdapQueryBuilder()
+                        .withSearchBase(searchBase)
+                        .withSearchFilter(searchFilter)
+                        .withAttributes(attributes)
+                        .build(),
+                searchResults -> {
+                    ImmutableSet.Builder<String> userAttributes = ImmutableSet.builder();
+                    while (searchResults.hasMore()) {
+                        final NamingEnumeration<? extends Attribute> attrs = searchResults.next().getAttributes().getAll();
+                        while (attrs.hasMore()) {
+                            final NamingEnumeration<?> avals = attrs.next().getAll();
+                            while (avals.hasMore()) {
+                                userAttributes.add(avals.next().toString());
+                            }
+                        }
+                    }
+                    return userAttributes.build();
+                });
+    }
+
+    public String lookupUserDistinguishedName(String userBaseDistinguishedName,
+            String searchFilter,
+            String bindDistinguishedName,
+            String bindPassword)
+            throws NamingException
+    {
+        Set<String> userDistinguishedNames = lookupUserDistinguishedNames(userBaseDistinguishedName,
+                searchFilter, bindDistinguishedName, bindPassword);
+        if (userDistinguishedNames.isEmpty()) {
+            if (allowUserNotExist) {
+                return "";
+            }
+            else {
+                String message = format("User filter [%s] failed for baseDN [%s]", searchFilter, userBaseDistinguishedName);
+                log.debug("%s", message);
+                throw new AccessDeniedException(message);
+            }
+        }
+        if (userDistinguishedNames.size() > 1) {
+            String message = format("Multiple group membership results for user [%s]: %s", userBaseDistinguishedName,
+                    userDistinguishedNames);
+            log.debug("%s", message);
+            throw new AccessDeniedException(message);
+        }
+        return getOnlyElement(userDistinguishedNames);
+    }
+
+    static String extractGroupName(
```

Comment:
Can we move the static methods to a Util class ?

Reply:
We could - but its analogous to the FileGroupProvider - which has static methods in it.